source/main/utils/handleCheckBlockReplayProgress.ts

```diff
@@ -65,15 +68,26 @@ const createHandleNewLogLine = (mainWindow: BrowserWindow) => {
       return;
     }
 
-    const progress = Math.floor(parseFloat(unparsedProgress));
+    // In rare cases cardano-node does not log 100%, therefore we need to manually mark the previous step as complete.
+    if (
+      type === BlockSyncType.replayedBlock &&
+      progressReport[BlockSyncType.validatingChunk] !== 100
+    ) {
+      progressReport[BlockSyncType.validatingChunk] = 100;
+    }
 
-    if (lastReportedProgressByType[type] !== progress) {
-      lastReportedProgressByType[type] = progress;
+    if (
+      type === BlockSyncType.pushingLedger &&
+      progressReport[BlockSyncType.replayedBlock] !== 100
+    ) {
+      progressReport[BlockSyncType.replayedBlock] = 100;
+    }
 
-      getBlockSyncProgressChannel.send(
-        { progress, type },
-        mainWindow.webContents
-      );
+    const progress = Math.floor(parseFloat(unparsedProgress));
+
+    if (progressReport[type] !== progress) {
+      progressReport[type] = progress;
+      getBlockSyncProgressChannel.send(progressReport, mainWindow.webContents);
```

Comment:
Instead of sending report for a specific type, let's send the whole progress report. This resolves the issue where the IPC channel subscription is registered late and messages from the main thread are sent into the void.

Question for people more experienced with IPC: is there any way to send a 'sticky' message (messsage which can be delivered to subscriptions registered after the message was sent)? If so, we could just slightly tweak the previous implementation which sends less data... However, I tried searching for it and couldn't find anything.

Reply:
![Uploading image.png…]()

Reply:
I don't know about the electron IPC implementation but the node IPC protocol itself does not support queueing messages. I think sending the entire `progressReport` object totally makes sense. It is not a concern from the performance perspective.

Reply:
Agreed sending a snapshot of `progressReport` , that's smart.